packages/core/content-manager/admin/src/features/ModelRBAC.tsx

```diff
@@ -0,0 +1,175 @@
+import * as React from 'react';
+
+import {
+  useRBAC,
+  useAuth,
+  type Permission,
+  createContext,
+  Page,
+  useQueryParams,
+} from '@strapi/admin/strapi-admin';
+
+import type { Schema } from '@strapi/types';
+
+/**
+ * The boolean values indicate the global actions a user can perform on the document.
+ * The `string[]` values tell us specifically which fields the actions can be performed on,
+ * for example, if the `canReadFields` array is empty, than no fields can be read by the user.
+ * This can happen even if the user can read the document.
+ */
+interface ModelRBACContextValue {
+  canCreate?: boolean;
+  canCreateFields: string[];
+  canDelete?: boolean;
+  canPublish?: boolean;
+  canRead?: boolean;
+  canReadFields: string[];
+  canUpdate?: boolean;
+  canUpdateFields: string[];
+  canUserAction: (
+    fieldName: string,
+    fieldsUserCanAction: string[],
+    fieldType: Schema.Attribute.Kind
+  ) => boolean;
+  isLoading: boolean;
+}
+
+const [ModelRBACProvider, useModelRBAC] = createContext<ModelRBACContextValue>('ModelRBAC', {
+  canCreate: false,
+  canCreateFields: [],
+  canDelete: false,
+  canPublish: false,
+  canRead: false,
+  canReadFields: [],
+  canUpdate: false,
+  canUpdateFields: [],
+  canUserAction: () => false,
+  isLoading: false,
+});
+
+interface ModelRBACProps {
+  children: React.ReactNode;
+  permissions: Permission[] | null;
+  model: string;
+}
+
+/**
+ * @internal This component is not meant to be used outside of the Content Manager plugin.
+ * It depends on knowing the model of the content-type using an argument passed.
+ * If you do use the hook outside of the context, we default to `false` for all actions.
+ *
+ * It then creates an list of `can{Action}` that are passed to the context for consumption
+ * within the app to enforce RBAC.
+ */
+const ModelRBAC = ({ children, permissions, model }: ModelRBACProps) => {
```

Comment:
As far as I can tell the only real difference between DocumentRBAC and this file is the additon of model prop and removal of `slug` from useParams.

Couldn't you avoid duplicating all this code by passing model as a prop, make it optional, if it's not passed fall back to the slug from useParams?

```tsx
const DocumentRBAC = ({ children, permissions, model }: DocumentRBACProps) => {
  const { slug: modelSlug } = useParams<{ slug: string }>();

  const slug = model ?? modelSlug

  if (!slug) {
    throw new Error('Cannot find the slug param in the URL');
  }
```

Reply:
i don't think that specific solution works because on the edit view we'll always get a param match from the URL, but it won't be the right one when dealing with a relation to another content type in a modal

Reply:
Hey, thanks for the suggestions, the PR is in draft and I duplicated the code only to have a working initial solution, not as a final solution. I wanted to understand from you first if the solution works as expected

Reply:
I removed the ModelRBAC and used the DocumentRBAC instead